sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+#include <string>
+#include <memory>
+
+namespace Azure { namespace Core {
+
+  //@param <T> The type of the final result of the long-running operation.
+  template <class T> class Operation {
+  private:
+    virtual std::unique_ptr<Http::RawResponse> PollInternal(Context& context) = 0;
```

Comment:
They are private methods in the class.  There is no accessibility to them so naming isn't dictated other than Pascal

Reply:
Why are they pure virtual then if they aren't mean to be accessible outside the class?

http://www.gotw.ca/publications/mill18.htm
> Guideline #2: Prefer to make virtual functions private.
> This lets the derived classes override the function to customize the behavior as needed, without further exposing the virtual functions directly by making them callable by derived classes (as would be possible if the functions were just protected). The point is that virtual functions exist to allow customization; unless they also need to be invoked directly from within derived classes' code, there's no need to ever make them anything but private. 

From https://stackoverflow.com/a/2170696/12509023

Also from https://stackoverflow.com/a/3978552/12509023
> The confusion is common enough, that C++ FAQ advocated against using private virtuals, for a long time, because confusion seemed to be a bad thing.

Though I can't find mention of that in the provided link. @CaseyCarter - thoughts?

Reply:
private & virtual doesn't make sense to me. Pure makes less sense as a derived class must implement it but it can't because it's private.
Maybe I'm missing something specific to C++ here, but C# wouldn't even compile this.

Reply:
Operation<T> is a template.  The derived type must provide an implementation for these two functions.  They are private b/c we do not want them directly called.  Instead it is the overloads on Operation<T> which will call the derived types methods.

Reply:
> The derived type must provide an implementation for these two functions. They are private b/c we do not want them directly called.

Right, gotcha. So, since they are not _actually_ private (i.e. we can't rename or break them), the original questions come back:
- Is this a common pattern to have private virtual methods suffixed with `Internal`? Should we name it differently or are we establishing a pattern here which we may want to add to the guidelines?
- Would we want such APIs to show up in APIView?

Also, I understand we want it to be virtual and private. Does it have to be **pure** virtual though?

Reply:
> The derived type must provide an implementation

Then we need it to be pure virtual right?  ` = 0`  otherwise overriding it is optional

Reply:
It should be protected, not private

Reply:
Making it private restricts even derived types from calling it, which has some value. They can override the behavior, but when and how it is called is controlled by the base type. That can be a useful mechanism to scope and isolate what the derived types can do.

The problem I see is, currently we aren't doing anything significant to control that behavior in some way (like checking context/etc.) and end up just calling their overridden method directly without any prologue/epilogue ceremony at the call site.

Reply:
> Making it private restricts even derived types from calling it

That's partially true.  The derived class can only call its own implementation.

> but when and how it is called is controlled by the base type.

No, the derived class can call its own implementation whenever it wants. You can have this:
```cpp
class Base {
private:
  virtual int Foo() = 0;
};

class Vh : public Base {
private:
  int Foo() override { return 1; }

public:
  // As derived class, I can call my implementation whenever I want
  int DoSomething() { return Foo(); }
};
```

The only thing different from making the pure virtual method `protected` and not `private` is to block derived classes from calling the virtual base function directly. But this would make no sense at all!
Like
```cpp
class Base {
protected:
  virtual int Foo() = 0;
};

class Vh : public Base {
private:
  int Foo() override { return 1; }

public:
  // There would have no sense to call the base to redirect to my own implementation
  int DoSomething() { return Base::Foo(); }
};
```

Reply:
Private virtual is a legit C++ concept called NVI (non-virtual interface), Herb Sutter is a huge fan.

Reply:
Thanks Anton. I found this: https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Non-Virtual_Interface
So now I see that this is legit C++ but I still don't fully understand. Why not just make the methods protected instead of private?